ethereum/contracts/ERC20App.sol

```diff
@@ -46,116 +53,69 @@ contract ERC20App is AccessControl {
         uint128 amount
     );
 
-    event Upgraded(
-        address upgrader,
-        Channel basic,
-        Channel incentivized
-    );
-
-    struct Channel {
-        address inbound;
-        address outbound;
-    }
-
-    bytes32 public constant INBOUND_CHANNEL_ROLE =
-        keccak256("INBOUND_CHANNEL_ROLE");
-
-    bytes32 public constant CHANNEL_UPGRADE_ROLE =
-        keccak256("CHANNEL_UPGRADE_ROLE");
-
-
-    constructor(Channel memory _basic, Channel memory _incentivized) {
-        Channel storage c1 = channels[ChannelId.Basic];
-        c1.inbound = _basic.inbound;
-        c1.outbound = _basic.outbound;
-
-        Channel storage c2 = channels[ChannelId.Incentivized];
-        c2.inbound = _incentivized.inbound;
-        c2.outbound = _incentivized.outbound;
-
-        _setupRole(CHANNEL_UPGRADE_ROLE, msg.sender);
-        _setRoleAdmin(INBOUND_CHANNEL_ROLE, CHANNEL_UPGRADE_ROLE);
-        _setRoleAdmin(CHANNEL_UPGRADE_ROLE, CHANNEL_UPGRADE_ROLE);
-        _setupRole(INBOUND_CHANNEL_ROLE, _basic.inbound);
-        _setupRole(INBOUND_CHANNEL_ROLE, _incentivized.inbound);
+    constructor(
+        address channelRegistry
+    ) {
+        registry = ChannelRegistry(channelRegistry);
     }
 
     function lock(
         address _token,
         bytes32 _recipient,
         uint128 _amount,
-        ChannelId _channelId,
         uint32 _paraID,
-        uint128 _fee
+        uint128 _fee,
+        uint32 _channelID
     ) public {
-        require(
-            _channelId == ChannelId.Basic ||
-                _channelId == ChannelId.Incentivized,
-            "Invalid channel ID"
-        );
+        address channel = registry.outboundChannelForID(_channelID);
+        if (channel == address(0)) {
+            revert UnknownChannel(_channelID);
+        }
 
+        if (!IERC20(_token).transferFrom(msg.sender, address(this), _amount)) {
+            revert InsufficientAllowance();
```

Comment:
`transferFrom` can fail with both an insufficient allowance but also insufficient balance. This error might hide the real error. 
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L237